ENH: Make group_var_ use Welford's algorithm.

jvkersch opened the pull request:
closes #10448 

This PR reimplements the Cython functions `group_var_float64` and `group_var_float32` to use Welford's algorithm, rather than the sum-of-squares method, which is numerically unstable. This came up in #10448; see also #10242 for more discussion.

Commit:
ENH: Make group_var_ use Welford's algorithm.

jvkersch commented:
@jreback Would you be willing to give this a review? I can probably tackle `_nanvar` at some point too, but not before the weekend.

jreback commented:
looks good. pls add:

- release note in 0.17.0 whatsnew (you can put in bug fix section)
- add a test that should match before this PR (to assure there isn't a calculation regression), e.g. something that isn't an edge case 
- squash

jvkersch commented:
@jreback Thanks for your feedback. I added some tests and squashed everything onto one commit. I also took the liberty of fixing up the "closes" trigger in the PR description: this addresses #10448 --- I have some ideas for #10242 but would like to address that in a separate PR.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -133,4 +133,7 @@ Bug Fixes
+
 - Bug in ``read_msgpack`` where input type is not always checked (:issue:`10369`)
+
+- Bug in `group_var_float64` which caused variance to be inaccurate (:issue:`10448`)
```
say something like ``groupby.var`` with small float values

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -285,6 +287,133 @@ def test_dropna(self):
+
+    _multiprocess_can_split_ = True
+
+    __test__ = False
```
what is this for?

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -285,6 +287,133 @@ def test_dropna(self):
             pd.Series([2, 1, 1], index=[5., 10.3, np.nan]))
 
+
+class TestGroupVar(tm.TestCase):
```
make this sub-class object, remove the ``__test__``

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -285,6 +287,133 @@ def test_dropna(self):
+        tm.assert_almost_equal(out[0, 0], 0.0)
+
+
+class TestGroupVarFloat64(TestGroupVar):
```
subclass ``(TestGroupVar, tm.TestCase)``

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -285,6 +287,133 @@ def test_dropna(self):
+        tm.assert_almost_equal(out[0, 0], 1.0 / 12, check_less_precise=True)
+
+
+class TestGroupVarFloat32(TestGroupVar):
```
same as above

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -285,6 +287,133 @@ def test_dropna(self):
+
+    algo = algos.algos.group_var_float32
+
+    dtype = np.float32
```
tighten up the whitespace, son't leave blank lines

jreback commented:
couple of style issues, otherwise looks good. ping when pushed / green

jvkersch commented:
@jreback Done and Travis shows green. Thanks for the review!

jreback commented:
thanks! this looks gr8!
